common/src/main/java/com/microsoft/identity/common/migration/DefaultSharedPrefsFileManagerReencrypter.java

```diff
@@ -0,0 +1,167 @@
+                shouldAbort
+        );
+
+        // Clear any keys marked for removal...
+        clearEntriesMarkedForRemoval(fileManager, cacheEntries, keysMarkedForRemoval);
+
+        if (shouldAbort.get()) {
+            return result;
+        }
+
+        // Write the newly encrypted records...
+        for (final Map.Entry<String, String> cacheEntry : cacheEntries.entrySet()) {
+            fileManager.putString(cacheEntry.getKey(), cacheEntry.getValue());
+        }
+
+        return result;
+    }
+
+    private interface Callable<T> {
+        void call(T t) throws Exception;
+    }
+
+    private void applyCacheMutation(@NonNull final Map<String, String> cacheEntries,
+                                    @NonNull final Callable<Map.Entry<String, String>> callable,
+                                    @NonNull final MigrationOperationResult inputResult,
+                                    @NonNull final ReencryptionParams params,
+                                    @NonNull final Set<String> keysMarkedForRemoval,
+                                    @NonNull final Set<String> skipKeys,
+                                    @NonNull AtomicBoolean shouldAbort) {
+        for (final Map.Entry<String, String> cacheEntry : cacheEntries.entrySet()) {
+            try {
+                if (skipKeys.contains(cacheEntry.getKey())) {
+                    continue;
+                }
+                callable.call(cacheEntry);
+            } catch (final Exception e) {
+                inputResult.addFailure(e);
+                skipKeys.add(cacheEntry.getKey());
+
+                if (params.eraseEntryOnError()) {
+                    keysMarkedForRemoval.add(cacheEntry.getKey());
+                }
+
+                if (params.eraseAllOnError()) {
+                    keysMarkedForRemoval.addAll(cacheEntries.keySet());
+                }
+
+                if (params.abortOnError()) {
+                    shouldAbort.set(true);
+                    break;
+                }
+            }
+        }
+    }
+
+    private void clearEntriesMarkedForRemoval(@NonNull final ISharedPreferencesFileManager fileManager,
+                                              @NonNull final Map<String, String> cacheEntries,
+                                              @NonNull final Set<String> keysMarkedForRemoval) {
+        for (final String removedKey : keysMarkedForRemoval) {
+            cacheEntries.remove(removedKey);
+            fileManager.remove(removedKey);
+        }
+    }
+
+    @Override
+    public void reencryptAsync(@NonNull final ISharedPreferencesFileManager fileManager,
+                               @NonNull final IStringEncrypter encrypter,
+                               @NonNull final IStringDecrypter decrypter,
+                               @NonNull final ReencryptionParams params,
+                               @NonNull final TaskCompletedCallback<IMigrationOperationResult> callback) {
+        new Thread(new Runnable() {
```

Comment:
I must have asked this before, but can we put this in a pool?

Reply:
Can do